modules/dnn/src/plugin_wrapper.impl.hpp

```diff
@@ -0,0 +1,319 @@
+            CV_LOG_ERROR(NULL, "DNN: wrong OpenCV major version used by plugin '" << api_header.api_description << "': " <<
+                cv::format("%d.%d, OpenCV version is '" CV_VERSION "'", api_header.opencv_version_major, api_header.opencv_version_minor))
+            return false;
+        }
+        if (!checkMinorOpenCVVersion)
+        {
+            // no checks for OpenCV minor version
+        }
+        else if (api_header.opencv_version_minor != CV_VERSION_MINOR)
+        {
+            CV_LOG_ERROR(NULL, "DNN: wrong OpenCV minor version used by plugin '" << api_header.api_description << "': " <<
+                cv::format("%d.%d, OpenCV version is '" CV_VERSION "'", api_header.opencv_version_major, api_header.opencv_version_minor))
+            return false;
+        }
+        CV_LOG_DEBUG(NULL, "DNN: initialized '" << api_header.api_description << "': built with "
+            << cv::format("OpenCV %d.%d (ABI/API = %d/%d)",
+                 api_header.opencv_version_major, api_header.opencv_version_minor,
+                 api_header.min_api_version, api_header.api_version)
+            << ", current OpenCV version is '" CV_VERSION "' (ABI/API = " << abi_version << "/" << api_version << ")"
+        );
+        if (api_header.min_api_version != abi_version)  // future: range can be here
+        {
+            // actually this should never happen due to checks in plugin's init() function
+            CV_LOG_ERROR(NULL, "DNN: plugin is not supported due to incompatible ABI = " << api_header.min_api_version);
+            return false;
+        }
+        if (api_header.api_version != api_version)
+        {
+            CV_LOG_INFO(NULL, "DNN: NOTE: plugin is supported, but there is API version mismath: "
+                << cv::format("plugin API level (%d) != OpenCV API level (%d)", api_header.api_version, api_version));
+            if (api_header.api_version < api_version)
+            {
+                CV_LOG_INFO(NULL, "DNN: NOTE: some functionality may be unavailable due to lack of support by plugin implementation");
+            }
+        }
+        return true;
+    }
+
+public:
+    std::shared_ptr<cv::plugin::impl::DynamicLib> lib_;
+    const OpenCV_DNN_Plugin_API* plugin_api_;
+
+    PluginDNNBackend(const std::shared_ptr<cv::plugin::impl::DynamicLib>& lib)
+        : lib_(lib)
+        , plugin_api_(NULL)
+    {
+        initPluginAPI();
+    }
+
+    std::shared_ptr<cv::dnn_backend::NetworkBackend> createNetworkBackend() const
+    {
+        CV_Assert(plugin_api_);
+
+        CvPluginDNNNetworkBackend instancePtr = NULL;
+
+        if (plugin_api_->v0.getInstance)
+        {
+            if (CV_ERROR_OK == plugin_api_->v0.getInstance(&instancePtr))
+            {
+                CV_Assert(instancePtr);
+                // TODO C++20 "aliasing constructor"
+                return std::shared_ptr<cv::dnn_backend::NetworkBackend>(instancePtr, [](cv::dnn_backend::NetworkBackend*){});  // empty deleter
```

Comment:
Maybe we could capture `shared_from_this()` inside a deleter for now?
Seems to achieve the same goal.

Reply:
1. This requires changes in class inheritance (use `std::enable_shared_from_this<T>`). I prefer to keep interfaces clear as much as possible.
2. I don't want to transfer object lifetime control from the plugin (it could not prevent unloading of plugin code anyway).

Reply:
1. Line 22: `class PluginDNNBackend CV_FINAL: public std::enable_shared_from_this<PluginDNNBackend>`
2. Then why `shared_ptr` is used in the first place?
3. Does TODO makes sense then?

Reply:
1. `PluginDNNBackend` != `NetworkBackend`.
2. To avoid uncontrolled usage of RAW pointers.
3. AFAIK, to alias on plugins instance (the same is for highgui/parallel plugins - code borrowed from there)

Reply:
1. We are in a class that is derived from enable_shared_from_this. The idea behind capturing shared_from_this()(PluginDNNBackend) inside a deleter of NetworkBackend was that the former wouldn't run its destructor before the latter.
2. In its current state they are equivalent.
3. The idea seems to be similar to p.1 above.

Reply:
I believe we need a separate proposal which covers similar cases from other plugins.